fe/fe-core/src/main/java/com/starrocks/sql/analyzer/Authorizer.java

```diff
@@ -278,39 +280,54 @@ public static void checkActionInDb(UserIdentity currentUser, Set<Long> roleIds,
      * Currently, it's used by `show databases` or `use database`.
      */
     public static void checkAnyActionOnOrInDb(UserIdentity currentUser, Set<Long> roleIds, String catalogName, String db)
             throws AccessDeniedException {
         Preconditions.checkNotNull(db, "db should not null");
+        AccessController controller = getInstance().getAccessControlOrDefault(catalogName);
 
+        // Check privilege on database
         try {
-            getInstance().getAccessControlOrDefault(catalogName).checkAnyActionOnDb(currentUser, roleIds, catalogName, db);
-        } catch (AccessDeniedException e1) {
+            controller.checkAnyActionOnDb(currentUser, roleIds, catalogName, db);
+            return;
+        } catch (AccessDeniedException e) {
+            if (!CatalogMgr.isInternalCatalog(catalogName)) {
+                throw new AccessDeniedException();
+            }
+        }
+
+        // Check privilege on other objects
+        List<AccessControlChecker> checkers = ImmutableList.of(
+                () -> controller.checkAnyActionOnAnyTable(currentUser, roleIds, catalogName, db),
+                () -> controller.checkAnyActionOnAnyView(currentUser, roleIds, db),
+                () -> controller.checkAnyActionOnAnyMaterializedView(currentUser, roleIds, db),
+                () -> controller.checkAnyActionOnAnyFunction(currentUser, roleIds, db),
+                () -> controller.checkAnyActionOnPipe(currentUser, roleIds, new PipeName("*", "*"))
+        );
+
+        AccessDeniedException lastExcepton = null;
+        for (AccessControlChecker checker : checkers) {
             try {
-                getInstance().getAccessControlOrDefault(catalogName)
-                        .checkAnyActionOnAnyTable(currentUser, roleIds, catalogName, db);
-            } catch (AccessDeniedException e2) {
-                if (CatalogMgr.isInternalCatalog(catalogName)) {
-                    try {
-                        getInstance().getAccessControlOrDefault(InternalCatalog.DEFAULT_INTERNAL_CATALOG_NAME)
-                                .checkAnyActionOnAnyView(currentUser, roleIds, db);
-                    } catch (AccessDeniedException e3) {
-                        try {
-                            getInstance().getAccessControlOrDefault(InternalCatalog.DEFAULT_INTERNAL_CATALOG_NAME)
-                                    .checkAnyActionOnAnyMaterializedView(currentUser, roleIds, db);
-                        } catch (AccessDeniedException e4) {
-                            getInstance().getAccessControlOrDefault(InternalCatalog.DEFAULT_INTERNAL_CATALOG_NAME)
-                                    .checkAnyActionOnAnyFunction(currentUser, roleIds, db);
-                        }
-                    }
-                } else {
-                    throw new AccessDeniedException();
-                }
+                checker.check();
```

Comment:
for external catalog, we don't need to check mv, function, pipe

Reply:
it's checked in line 305

Reply:
table need to be checked for external catalog

Reply:
fixed